### PR TITLE
Parse mvex to retrieve fragment_duration.

### PR DESCRIPTION
**Pull request by alfredoyang**
Parse fragment_duration for https://bugzilla.mozilla.org/show_bug.cgi?id=1304254.

Test will be added in another PR later.

### CODE REVIEW

File: mp4parse/src/lib.rs [hunk 1]
```diff
@@ -565,6 +571,32 @@ fn read_moov<T: Read>(f: &mut BMFFBox<T>, context: &mut MediaContext) -> Result<
     Ok(())
 }
 
+fn read_mvex<T: Read>(src: &mut BMFFBox<T>) -> Result<MovieExtendsBox> {
+    let mut iter = src.box_iter();
+    let mut duration: Option<u64> = None;
```
Comment (rillian) [inline mp4parse/src/lib.rs]: The compiler should be able to infer this type from the signature of `read_mehd` so you only need this if you especially want to document it locally. Up to you.

File: mp4parse/src/lib.rs [hunk 2]
```diff
@@ -565,6 +571,32 @@ fn read_moov<T: Read>(f: &mut BMFFBox<T>, context: &mut MediaContext) -> Result<
...
+            },
+            _ => try!(skip_box_content(&mut b)),
+        }
+    }
+    Ok(MovieExtendsBox {
+        fragment_duration: duration.map_or(None, |x| { Some(MediaScaledTime(x)) }),
+    })
+}
+
+fn read_mehd<T: Read>(src: &mut BMFFBox<T>) -> Result<Option<u64>> {
```
Comment (rillian) [inline mp4parse/src/lib.rs]: `mehd` always has a fragment duration, so you don't need to return an Option here. A simpler `Result<u64>` or `Result<MediaScaledTime>` would work. This complicates the way you update duration, but I think it's better to have a temporary there and keep the semantics of the function signature correct.

e.g.
```rust
let mut fragment_duration = None;
...
            let duration = try!(read_mehd(&mut b));
            fragment_duration = Some(duration);
...
Ok(MovieExtendsBox {
    fragment_duration: fragment_duration,
})
```

File: mp4parse_capi/src/lib.rs [hunk 3]
```diff
@@ -503,8 +536,9 @@ pub unsafe extern fn mp4parse_is_fragmented(parser: *mut mp4parse_parser, track_
     let tracks = &context.tracks;
     (*fragmented) = false as u8;
 
-    if !context.has_mvex {
-        return MP4PARSE_OK;
+    match context.mvex {
+        Some(_) => {},
+        None => return MP4PARSE_OK,
```
Comment (alfredoyang) [inline mp4parse_capi/src/lib.rs]: Fixed.

File: mp4parse/src/lib.rs (hunk 1 above)
Comment (alfredoyang) [inline mp4parse/src/lib.rs]: Fixed.

File: mp4parse/src/lib.rs (hunk 2 above)
Comment (alfredoyang) [inline mp4parse/src/lib.rs]: Fixed.

**Comment by alfredoyang** [comment]
Should I update parser to gecko in the same bug? Because there is another gecko patch depending on this in the same bug.

**Comment by rillian** [comment]
If you need these changes in gecko we can bump the in-tree version on the bugzilla side. You can do it, or I can, I've had lots of practice.

**Pull request merged**